src/net_processing.cpp

```diff
@@ -3969,11 +3969,58 @@ void PeerManager::ConsiderEviction(CNode& pto, int64_t time_in_seconds)
 
 void PeerManager::EvictExtraOutboundPeers(int64_t time_in_seconds)
 {
-    // Check whether we have too many outbound peers
-    int extra_peers = m_connman.GetExtraOutboundCount();
+    // If we have any extra block-relay-only peers, disconnect the youngest unless
+    // it's given us a block -- in which case, compare with the second-youngest, and
+    // out of those two, disconnect the peer who least recently gave us a block.
+    // The youngest block-relay-only peer would be the extra peer we connected
+    // to temporarily in order to sync our tip; see net.cpp.
+    // Note that we use higher nodeid as a measure for most recent connection.
+    int extra_block_relay_peers = m_connman.GetExtraBlockRelayCount();
+    if (extra_block_relay_peers > 0) {
+        std::pair<NodeId, int64_t> youngest_peer = {-1, 0};
+        std::pair<NodeId, int64_t> next_youngest_peer = {-1, 0};
+
+        m_connman.ForEachNode([&](CNode* pnode) {
+            LockAssertion lock(::cs_main);
```

Comment:
The two lock assertions added in this function need to be updated to build on current master:
```diff
@@ -3964,8 +3964,8 @@ void PeerManager::EvictExtraOutboundPeers(int64_t time_in_seconds)
         std::pair<NodeId, int64_t> youngest_peer = {-1, 0};
         std::pair<NodeId, int64_t> next_youngest_peer = {-1, 0};
 
-        m_connman.ForEachNode([&](CNode* pnode) {
-            LockAssertion lock(::cs_main);
+        m_connman.ForEachNode([&](CNode* pnode) EXCLUSIVE_LOCKS_REQUIRED(::cs_main) {
+            AssertLockHeld(::cs_main);
             if (!pnode->IsBlockOnlyConn() || pnode->fDisconnect) return;
@@ -3979,8 +3979,8 @@ void PeerManager::EvictExtraOutboundPeers(int64_t time_in_seconds)
             // disconnect our second youngest.
             to_disconnect = next_youngest_peer.first;
         }
-        m_connman.ForNode(to_disconnect, [&](CNode* pnode) {
-            LockAssertion lock(::cs_main);
+        m_connman.ForNode(to_disconnect, [&](CNode* pnode) EXCLUSIVE_LOCKS_REQUIRED(::cs_main) {
+            AssertLockHeld(::cs_main);
             // Make sure we're not getting a ```

Reply:
(see #19979)

Reply:
Thanks, updated.